Lots of Tweaks and Functionality Added

nichogenius opened the pull request:
Sorry for the repeat request.  I put a few more hours of work into this today since my last pull request, so I figured it's probably better to do it in one pull, especially as my last pull request had a couple of bugs.

Honestly, I've made so many changes that it's probably easier to try the new version rather than review every single line I've changed.  I don't believe I removed or modified any existing functionality (other than maybe 3 patterns that were giving lots of false positives).  

--New Features
-patterns_iraw.txt - case insensitive version of patterns_raw.txt
-comments and white space in pattern files
-A short flag option for every long flag option and visa-versa
-Re-formatted output
  MD5 hash is now listed before the path and is optional. Makes pasting into the whitelist easier
  file changed time is also optional
  comments on what the pattern means are also optional
  Added '#' characters to output to make the output 'bash shell safe' to avoid accidental code executions 
  should the script output be dumped into a terminal from the clipboard
  Added a new 'blu-ish' color to visually isolate the path from the hash and time data

--New Flags
-b/--base64 - Uses two large pattern files to look for PHP keywords and functions encoded in base64.  Normal pattern files are ignored in this mode.
-m/--checksum - md5 value is now hidden by default, but can be shown with this flag
-c/--comment - shows the most recent comment line from the pattern file to appear before the matched pattern.  Spammy, but more informative, especially in base64 mode.
-n/--no-color - sets the color strings to an empty string which prevents them from being displayed in the output. Useful for dumping output to a file.
-t/--time -shows the last changed time of the inode, or unix ctime.
-v/--verbose -Don't stop scanning a file after the first match is found.  Display all matches.  Slower, but more informative.  Especially useful in base64 mode.

--New Files
-php_functions.txt & php_keywords.txt - Used by -b/--base64 flag instead of patterns_raw.txt.  Useful as just a secondary scan option and as a way to find new common base64 patterns in known malware.
-b64p.py - Python script that converts a plain text string into the 3 base64 equivalents.

--Lots of general code maintenance
-Added comments and structure to the pattern files... organizing them will take more work.
-Added comments to the class functions
-Added lots of base64 strings to the patterns_raw.txt file
-Added some keywords to patterns_iraw.txt which often catch phishing activity (ie: 'phishtank')
-Split up and re-vised many functions
-Pattern files are now loaded only once at startup, and not re-loaded for each file scanned
-scan() function split up to a more modular version
-out() function replaced with error()
-added printPath() function for the result printing
-added parseArgs() function to handle all option reading
-Alphabetized the function definitions in scan.php
-Sorted the whitelist file first by path, then by hash/checksum

nichogenius committed:
Sorted this file by column 2

cat whitelist.txt | sort -k 2 | less
No reason this shouldn't be sorted perfectly to keep like files together.
No white list rules changed... just plain sorting.

nichogenius committed:
Two-key sorting is better

cat php-malware-scanner-master/whitelist.txt | sort -k 2,2 -k 1,1 | less
More of an OCD thing than anything, but might as well sort primarily by file path, secondarily by hash value.

nichogenius committed:
Whitespace and Comment handling for pattern files

The pattern files are large and complex enough to justify some whitespace and comments to explain what each entry is.

Added logic to check if the line is empty or if the first character is equal to '#' before using it as a pattern.  Simply skips over empty and commented lines.

nichogenius committed:
Added generic base64 long string pattern

Added some comments to the file and added a generic base64 string regex for long base64 strings.

nichogenius committed:
Organizing, categorizing and prioritizing patterns

There's enough raw patterns in here to justify organizing the file.
Now that whitespace and comments are supported, I've been dividing it into sections

More critical problems should be near the top as I would rather the script identify a file as a backdoor instead of as a spammer.

I don't know the history behind a lot of these or the implication of the code, so I'm sure I mis-categorized many.  There are also many that I have not done yet.

nichogenius committed:
Bug fix for last commit

comment and empty line filters were in the wrong place

nichogenius committed:
using strpos instead of substr_count

don't know if it's faster, but I don't see a reason to count the number of times a line exists in a file for our use case.

nichogenius committed:
New file for case-insensitive search of strings

New file for case-insensitive search of strings
patterns_iraw.txt

nichogenius committed:
Added case-insensitive search logic

Added case-insensitive search logic

will search patterns_raw.txt, patterns_iraw.txt and patterns_re.txt

nichogenius committed:
Added Security service providers

Phishers often block these web-security providers... worth checking for if they don't create many false positives.

nichogenius committed:
Add files via upload

This is a file of base64 patterns that represent strings that would be present if any of the functions in php7 were encoded to base64.  I'll probably add structure later by grouping them with their plain text translation.

This file is useful to swap out with patterns_raw.txt to gain additional insights into other strings to search for in base64.

nichogenius committed:
Removed some nasty false positive patterns

Removed some nasty false positive patterns

nichogenius committed:
Added functions, dropped Spammer.

Spammers gives false positives.  added a couple more php functions in base64

nichogenius committed:
Removed (chr\(\d+\)\.){4,}

Too many false positives

nichogenius committed:
Adding base_convert as a test

base_convert seems like a rare enough function... trying it out to see if it generates many fp's.

nichogenius committed:
bad idea... lots of fps

bad idea... lots of fps
removing base_convert

nichogenius committed:
ovh removed

ovh removed

nichogenius committed:
Uploaded an organized and commented version

This file just contains a list of internal php 7 functions (probably incomplete depending on extensions etc) and their 3 base64 fingerprints. It is designed to be used as either a pattern file to explore potential patterns that may be effective, or simply as a reference to translate between plain text php and the 3 different base64 versions.

nichogenius committed:
just adding some comments

just adding some comments

nichogenius committed:
added create_function

interesting note from the php.net manual on create_function:
Caution
This function internally performs an eval() and as such has the same security issues as eval(). Additionally it has bad performance and memory usage characteristics.

nichogenius committed:
added file_get_contents, fopen

dropped allow_url_fopen as fopen will catch those references on its own.

nichogenius committed:
shortened base64_decode to just base64

shortened the base64 fingerprints of 'base64_decode' to just 'base64'. will also catch cases of base64_encode which isn't quite so bad but still worth finding.

nichogenius committed:
Typo fix

Typo fix

nichogenius committed:
Added obfuscat

obfuscat should catch things like obfuscated, obfuscator, obfuscation.
Some online obfuscator providers leave a signature comment in the code.

nichogenius committed:
Added http://www.fopo.com.ar/

http://www.fopo.com.ar/ is a free online php obfuscator that apparently leaves comments in the code which we should be looking for.

nichogenius committed:
removed datatran

removed datatran for false positives

nichogenius committed:
obfuscat is too common, causes fp's

obfuscat is too common, causes fp's

nichogenius committed:
Added createfunction and gzinflate obfuscations

Added createfunction and gzinflate obfuscations.
also added gzinflate b64 patterns

nichogenius committed:
Found a bug in my base64 converter

Found a bug in my base64 converter

My base64 conversion script is supposed to find the maximum length string that is guaranteed to be present if the input plain text string is somewhere in the original plain text code, however there was an off by 1 error which made some patterns 1 character longer than they should have been.   Short patterns (ie 4 chars) were prone to false positives because they really were 3 character patterns which is too short to be useful.  Long patterns were likely missing results.

Should be fixed now.

nichogenius committed:
This is how I generate base64 sample patterns.

Example usage:

I want to see if a giant block of base64 code contains any references to the string 'base64'. 
The naive approach is to convert the string to it's base64 equivalent, YmFzZTY0.

There are two problems with this approach.  The first is that the string will be different depending on the position of the first character 'Y' in the input string.  Possible offents are 0 bits, 2 bits or 4 bits.  The above example only calculates the 0 bit offset.  There should be 3 separate base64 strings to look for.

The second problem is that base64 strings use a 6 bit encoding, so the characters don't align the same as 8 bit encoding.  This leads to character bleeding at the beginning and ends of a string where the string will change depending on its immediate context.  This script calculates the maximum constant string length that should be present.  Unfortunately it requires trimming characters which can often lead to very short strings.

nichogenius committed:
I was wrong... original code was correct.

I was wrong... original code was correct... there was no bug but the one in my head

nichogenius committed:
this is not a bug... just fixing a bad bug fix

this is not a bug... just fixing a bad bug fix.
there is no off by 1 error after all.

nichogenius committed:
removed mail b64, added chr b64

removed mail b64, added chr b64

mail was generating too many false positives.
chr has only one pattern that is long enough to use with any sort of reliability, but it is one that we want to look out for anyway.

nichogenius committed:
Added array_ and cslashes

Found a couple of cases where the php functions array_shift and addcslashes were used in base64 encoded malware.

Adding strings to catch any references to 'cslashes' which will catch both addcslashes and strip cslashes
Adding strings to catch any references to 'array_' which will catch about a dozen array modification functions.

nichogenius committed:
PHP Keywords in Base64, pattern file

This is a file containing a list of PHP keywords converted to base64.  It's designed to be used as a pattern file to identify common keywords used in obfuscated code.

nichogenius committed:
Renamed to match naming conventionsi

Renamed to match naming conventionsi

nichogenius committed:
Copied comments from php_functions.php

Copied comments from php_functions.php

nichogenius committed:
Adding str_, function, echo and include in base64

str_ will match 13 separate php functions, many of which can be used for string/modifcation aka obfuscation
function added to catch function defining.
echo added as it is a  common php  keyword, though experimental... may cause a of false positives
include added as it is often used to link in other malware files.

nichogenius committed:
Added encoded versions of '_' character.

Added encoded versions of '_' character.

nichogenius committed:
Added new flag options

Added a single short flag for every long flag and a single long flag for every short flag.
This now gives us 2 ways to set each flag.
Also updated the showhelp.
Dropped an unnecessary 'else' statement.

nichogenius committed:
Added long single line PHP code pattern

One common tactic is to shove all of your PHP code into a single line, often contained within its own PHP tags, and drop it into any .php file that you want.  This pattern should detect if more than 750 characters are contained within PHP tags on a single line.

nichogenius committed:
preg_match 's' flag changed to 'm'

the 's' flag tells preg_match to operate in multi-line mode.  the 'm' flag does the same, but allows line begin and ends to still be matched which is useful in some cases.

nichogenius committed:
Added Experimental Patterns

Removed anyresults.net from the base64 pattern strings.

Added base64 patterns for $_POST $_GET $_REQUEST $_COOKIE extract and GLOBALS

nichogenius committed:
Added New Malware Signatures/Fingerprints

IndoXploit
FaisaL Ahmed aka rEd X
'F'.'il'.'esMan'
FilesMan

nichogenius committed:
Several Significant changes to scan.php

- Gave each flag option a short or long option; like i:ignore or d:directory or k:hide-ok
- Added a verbose option that instructs the scan to scan a file for ALL matches and not just stop at the first one.
- Restructured the output code to allow for the verbose flag, mainly a new function printPath and where the md5 hash is computed
- Modified the output to be cleaner, checksum is printed first as it is fix-width and to make it easier to paste into the whitelist file.
- Modified the output to be 'bash safe', ie when I accidentally paste my scan results into my terminal, the '#' should make sure everything is treated as a comment.  This is in contrast to possibly attempting to execute absolute paths to potentially malicious PHP scripts and the usage of the '>' which tells the shell to write to a file.  Also enclosed each path in {} for similar purposes.
- Printing the matched string/pattern in $color... might change later depending on  preference.

nichogenius committed:
Raised Long PHP line Limit

Previous limit of 750 chars enclosed in php tags in a single line was too low... false positives were being triggered by a w3 total cache file because some guy decided to print one gigantic long message in a single line.

Raising to 1100

nichogenius committed:
Removing fopen('/etc/passwd - Redundant

fopen('/etc/passwd dropped
sub-pattern /etc/passwd exists and is completely contained within fopen('/etc/passwd

nichogenius committed:
FilesMan Needs to be a regex

FilesMan Needs to be a regex

nichogenius committed:
Verbose Bug fix and pattern loading optimization

Verbose flag was not proceeding with the next scan due to !found being set.  Added a check to see if it is verbose when it decides to do the next scan.

Patterns should be loaded once and only once.  The files aren't large so not a problem with memory, however it might impact performance if we are loading the same 3 files ever time we scan a file.

nichogenius committed:
Pattern Loading Moved To Constructor

It makes more sense to put the one time pattern load code into the constructor rather than the scan method.

nichogenius committed:
Cleaned up pattern whitespace and comment handling

checking for comments and whitespace lines in pattern files should only be done once when the patterns are loaded.
Added this code to the loadPatterns() function and removed it from the 3 scan loops.

nichogenius committed:
Bug Fixes, added time/checksum flags, organized

--Fixed a bug with the out function.  Previous updates of mine did not update all calls to the out function which I changed the parameters for.  Fixed this by replacing the out function with an 'error' function.
--Alphabetized function definitions and did some general tidying up
--Made all functions private except the constructor.
--Created parseArgs function to handle reading in options.
--Fixed a bug with 'extra-check' where htaccess and googleBot were being pushed to the pattern array each time a file was scanned. 
 This bug was created when I moved the pattern initialize code to the constructor.  Moved extra-check code with the rest of the initialize pattern calls.
--Added -no-color, -time, and -checksum flags.  I'd prefer if the output was only as spammy as the user requests.  Time should be helpful in tracing when the attack occurred and if files are related to the same hack.  Time and checksum do not display by default.  no-color flag makes it easier to dump to plain text files.

nichogenius committed:
base64 pattern updates

nichogenius committed:
removed redundant pattern

'create_function' is redundant as 'function' is already present.

nichogenius committed:
added b64 pattern for 'require'

added b64 pattern for 'require'

nichogenius committed:
Added comment lines for each regex

Added comment lines for each regex
This is to allow the -comment flag to provide either an accurate comment or no comment at all.

nichogenius committed:
large 'scan' function broken up comment flag added

scan function has been broken up into a multiple functions which should make it more modular if future scan types are needed.
--comment, -a flag was added which prints the first comment to appear prior to the matched pattern in the pattern file.

nichogenius committed:
updated flags

updated flags

checksum can be -m for md5
comment can be -c for comment

nichogenius committed:
fixed -m checksum flag

fixed -m checksum flag

nichogenius committed:
Added Comments and Documentation

Added Comments and Documentation

nichogenius committed:
added a forgotten flag to the help

looks like i forgot to add the --no-color option to the help display.

nichogenius committed:
2 typos = 1 fixed bug

typo in showHelp - ignore spelled incorrectly
typo in pattern file load for the b64 keyword file... broke part of base64 mode.

scr34m commented:
Test run on the sample database with this PR was a bit more accurate, no problems found int matching.

What is the idea behind verbose flag, because it doesn't stop scanning at the first match and the naming is misleading, --no-stop or --continue is better?

Only a few modifiecations needed:
- update README.md with the new flags and new "patterns_iraw.txt" file
- move b64p.py into a tools directory

nichogenius committed:
Updated with new Usage Information

Updated with new Usage Information - Mostly just new flags.

nichogenius committed:
Moving into subdirectory 'tools' + renamed

nichogenius committed:
Moved into tools directory

Moved into tools directory

nichogenius committed:
Moved to base64_patterns

Moved to base64_patterns

nichogenius committed:
Moved to base64_patterns folder

Moved to base64_patterns folder

nichogenius committed:
moved to definitions

moved to definitions

nichogenius committed:
moved to definitions

moved to definitions

nichogenius committed:
moved to definitions

moved to definitions

nichogenius committed:
Updated flags, pattern paths

added
--all-output
--pattern
--no-stop

removed --verbose/-v

put color in the pattern comments
updated paths to pattern files to either base64_patterns, or definitions

nichogenius committed:
Bug fix - forgot '/' in the new path 

Bug fix - forgot '/' in the new path
patterns weren't loading :(

nichogenius commented:
README.md - updated
b64.py renamed to text2base64.py and placed in tools
php_functions.txt and php_keywords.txt placed in base64_patterns folder
patterns_raw.txt, patterns_iraw.txt and patterns_re.txt placed in definitions folder.

--verbose changed to --no-stop with -s flag
--pattern/-p flag added to show patterns
--all-output/-a flag added as a shortcut to -cmpt or --comment --checksum --pattern --time

nichogenius committed:
Updated with patterns_iraw.txt and text2base64.py

nichogenius committed:
Update README.md

nichogenius committed:
Update README.md

nichogenius committed:
Update README.md

nichogenius committed:
Update README.md

nichogenius committed:
Update README.md

nichogenius commented:
Updated README.md with additions.

nichogenius commented:
Are there any features that you've been planning to add that I can look into?

scr34m commented:
At the moment I don't have any idea to implement, thank you for the changes and contribution!
